### PR TITLE
Close #103: Add RTTI to words

### LINKED ISSUE #103: Consider addition RTTI to words
## Overview

Now RTTI (runtime type information) becomes obvious. It's required by .NET transaction support, by debug purposes like stack and head printing and transaction introspection. Also it will remove necessarity of type-bounded arithmetic instructions. I see two approaches to implementation.

1. Use [BSON](http://bsonspec.org/) or [MessagePack](https://msgpack.org). Pros: it's are well-known standards, stable third-party implementations. Cons: this is _serialization_ formats, both don't support references, hard to mutate.
2. More domain-specific format. Pros: we can use advantages of our VM without compromises. Cons: need to write a lot of code, bugs are coming, NIH is a shame.

I'm inclined to the second approach.

## Proposal

This is specification pseudo-BNF of word format.

```
length := 0b00<6 bits of data>
       | 0b01<6 bits length>
       | 0b10<14 bits length>
       | 0b11<22 bits length>

bytes := length byte[&length]

int8    := 0x01
int16   := 0x02
int32   := 0x03
bigint  := 0x04
uint8   := 0x05
uint16  := 0x06
uint32  := 0x07
decimal := 0x08
boolean := 0x09
ref     := 0x0A
utf8    := 0x0B
array   := 0x0C
struct  := 0x0D

primitive_type := int8
               | int16
               | int32
               | bigint
               | uint8
               | uint16
               | uint256
               | double
               | boolean
               | ref

type := primitive_type
      | struct
      | array
      | utf8

int8_data    := bytes~1
int16_data   := bytes~2
int32_data   := bytes~4
bigint_data   := length bytes[&length]
uint8_data   := bytes~1
uint16_data  := bytes~2
uint32_data  := bytes~4
double_data := bytes~8 # strict IEEE-754 floating point number
ref_data     := bytes~4
boolean_data := bytes~1

primitive_data := int8 int8_data
                | int16 int16_data
                | int32 int32_data
                | bitint bigint_data
                | uint8 uint8_data
                | uint16 uint16_data
                | uint32 uint32_data
                | double double_data
                | ref ref_data
                | boolean

data := primitive_data
      | array primitive_type length data(primitive_type)[&length]
      | struct bytes length (bytes, primitive_data)[&length] # struct_name, [(field, field_value)]
      | utf8 bytes
```

`smth[num]` means that we duplicate `smth` structure `num` times. `byte[8]` means 8 bytes, 
`bytes~num` means that we _expect_ `num` of bytes (which length is dynamic).

`&length` refers to given `length` field and means an integer representation of that field. 

`(a, b)` means pair type, e.g. two values of `a` and `b` are written consecutively. 

`data(primitive_type)` means corresponding `*_data` structure for `primtive_type`

### PR DESCRIPTION
**Pull request by fomkin**
1. New `pravda.vm.state.Data` class with runtime for data and serialization/deserialization methods.
2. New `asm` with comments support, and valid assembler test rendering.
4. `forth` removed.
5. VM tests removed (NO TEST COVERAGE FOR NOW).

### CODE REVIEW

File: vm/src/main/scala/pravda/vm/DataOperations.scala [hunk 1]
```diff
@@ -0,0 +1,701 @@
...
+      case Uint8(rhs) => Int32(lhs + rhs)
+      case Uint16(rhs) => Int32(lhs + rhs)
+      case Uint32(rhs) => Int32((lhs + rhs).toInt)
+      case BigInt(rhs) => Int32((lhs + rhs).toInt)
+      case Number(rhs) => Int32((lhs + rhs).toInt)
+      case _ => throw VmErrorException(WrongType)
+    }
+    case Int16(lhs) => b match {
+      case Int8(rhs) => Int16((lhs + rhs).toShort)
+      case Int16(rhs) => Int16((lhs + rhs).toShort)
```
Comment (vovapolu) [inline vm/src/main/scala/pravda/vm/DataOperations.scala]: Why it doesn't get wider?

File: vm/src/main/scala/pravda/vm/DataOperations.scala [hunk 2]
```diff
@@ -0,0 +1,701 @@
...
+      case Int32(rhs) => Int16((lhs + rhs).toShort)
+      case Uint8(rhs) => Int16((lhs + rhs).toShort)
+      case Uint16(rhs) => Int16((lhs + rhs).toShort)
+      case Uint32(rhs) => Int16((lhs + rhs).toShort)
+      case BigInt(rhs) => Int16((lhs.toInt + rhs).toShort)
+      case Number(rhs) => Int16((lhs + rhs).toShort)
+      case _ => throw VmErrorException(WrongType)
+    }
+    case Int8(lhs) => b match {
+      case Int8(rhs) => Int32(lhs + rhs)
```
Comment (vovapolu) [inline vm/src/main/scala/pravda/vm/DataOperations.scala]: But here it gets wider somehow.

**Comment by vovapolu** [comment]
I don't quite understand the number widering in addition, multiplication and etc. And why did you delete all the vm tests? Are you going to reimplement them with another approach?

File: vm/src/main/scala/pravda/vm/DataOperations.scala (hunk 2 above)
Comment (fomkin) [reply vm/src/main/scala/pravda/vm/DataOperations.scala]: It was changed on my local machine. Now all operations save `lhs` type.

File: vm-api/src/main/scala/pravda/vm/Data.scala [hunk 3]
```diff
@@ -0,0 +1,641 @@
...
+    val (primitive, all, utf8) = {
+      import fastparse.all._
+
+      val ws = P(CharIn(Seq(' ', '\t', '\n', '\r')).rep)
+      val comma = P(ws ~ "," ~ ws)
+      val decDigs = P(CharIn('0' to '9').rep(1))
+      val hexDig = P(CharIn('0' to '9', 'a' to 'f', 'A' to 'F'))
+      val hexDigs = P(hexDig.rep(1))
+
+      val float = P("-".!.? ~ decDigs.! ~ "." ~ decDigs.! ~ (CharIn("eE") ~ CharIn("+-").? ~ decDigs.rep).!.?) map {
```
[the remark this answers is not in the document]
Comment (fomkin) [reply vm-api/src/main/scala/pravda/vm/Data.scala]: Well. We are not JSON. Just write `1.0e-2`.

File: vm-asm/src/main/scala/pravda/vm/asm/Operation.scala [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  def mnemonic: String
+}
+
+object Operation {
+
+  import Opcodes._
+
+  sealed abstract class ParametrizedOperation(val mnemonic: String) extends Operation
+
+  // Virtual operations (which don't included to bytecode)
```
[the remark this answers is not in the document]
Comment (fomkin) [reply vm-asm/src/main/scala/pravda/vm/asm/Operation.scala]: thx

File: vm-asm/src/main/scala/pravda/vm/asm/PravdaAssembler.scala [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+    var label = Option.empty[String]
+    while (buffer.hasRemaining) {
+      val opcode = buffer.get & 0xff
+      operations += Operation.operationByCode.getOrElse(
+        key = opcode,
+        default = (opcode: @switch) match {
+          case Opcodes.META =>
+            Metadata.readFromByteBuffer(buffer) match {
+              case Metadata.LabelDef(name) => Operation.Label(name)
+              case Metadata.LabelUse(name) =>
```
[the remark this answers is not in the document]
Comment (fomkin) [reply vm-asm/src/main/scala/pravda/vm/asm/PravdaAssembler.scala]: This way we mark what push+jump leads to the label.

**Pull request merged**